query_base.go

```diff
@@ -1096,3 +1096,235 @@ func (q cascadeQuery) appendCascade(fmter schema.Formatter, b []byte) []byte {
+	for _, idx := range indexes {
+		hints = append(hints, schema.UnsafeIdent(idx))
+	}
+	return hints
+}
+
+func (ih *idxHintsQuery) addUseIndex(indexes ...string) {
+	if len(indexes) == 0 {
+		return
+	}
+	ih.lazyUse().names = ih.appendIndexes(ih.use.names, indexes...)
+}
+
+func (ih *idxHintsQuery) addUseIndexForJoin(indexes ...string) {
+	if len(indexes) == 0 {
+		return
+	}
+	ih.lazyUse().forJoin = ih.appendIndexes(ih.use.forJoin, indexes...)
+}
+
+func (ih *idxHintsQuery) addUseIndexForOrderBy(indexes ...string) {
+	if len(indexes) == 0 {
+		return
+	}
+	ih.lazyUse().forOrderBy = ih.appendIndexes(ih.use.forOrderBy, indexes...)
+}
+
+func (ih *idxHintsQuery) addUseIndexForGroupBy(indexes ...string) {
+	if len(indexes) == 0 {
+		return
+	}
+	ih.lazyUse().forGroupBy = ih.appendIndexes(ih.use.forGroupBy, indexes...)
+}
+
+func (ih *idxHintsQuery) addIgnoreIndex(indexes ...string) {
+	if len(indexes) == 0 {
+		return
+	}
+	ih.lazyIgnore().names = ih.appendIndexes(ih.ignore.names, indexes...)
+}
+
+func (ih *idxHintsQuery) addIgnoreIndexForJoin(indexes ...string) {
+	if len(indexes) == 0 {
+		return
+	}
+	ih.lazyIgnore().forJoin = ih.appendIndexes(ih.ignore.forJoin, indexes...)
+}
+
+func (ih *idxHintsQuery) addIgnoreIndexForOrderBy(indexes ...string) {
+	if len(indexes) == 0 {
+		return
+	}
+	ih.lazyIgnore().forOrderBy = ih.appendIndexes(ih.ignore.forOrderBy, indexes...)
+}
+
+func (ih *idxHintsQuery) addIgnoreIndexForGroupBy(indexes ...string) {
+	if len(indexes) == 0 {
+		return
+	}
+	ih.lazyIgnore().forGroupBy = ih.appendIndexes(ih.ignore.forGroupBy, indexes...)
+}
+
+func (ih *idxHintsQuery) addForceIndex(indexes ...string) {
+	if len(indexes) == 0 {
+		return
+	}
+	ih.lazyForce().names = ih.appendIndexes(ih.force.names, indexes...)
+}
+
+func (ih *idxHintsQuery) addForceIndexForJoin(indexes ...string) {
+	if len(indexes) == 0 {
+		return
+	}
+	ih.lazyForce().forJoin = ih.appendIndexes(ih.force.forJoin, indexes...)
+}
+
+func (ih *idxHintsQuery) addForceIndexForOrderBy(indexes ...string) {
+	if len(indexes) == 0 {
+		return
+	}
+	ih.lazyForce().forOrderBy = ih.appendIndexes(ih.force.forOrderBy, indexes...)
+}
+
+func (ih *idxHintsQuery) addForceIndexForGroupBy(indexes ...string) {
+	if len(indexes) == 0 {
+		return
+	}
+	ih.lazyForce().forGroupBy = ih.appendIndexes(ih.force.forGroupBy, indexes...)
+}
+
+func (ih *idxHintsQuery) appendIndexHints(
+	fmter schema.Formatter, b []byte,
+) ([]byte, error) {
+	type IdxHint struct {
+		Name   string
+		Values []schema.QueryWithArgs
+	}
+
+	var hints []IdxHint
+	if ih.use != nil {
+		hints = append(hints, []IdxHint{
+			{
+				Name:   "USE INDEX",
+				Values: ih.use.names,
+			},
+			{
+				Name:   "USE INDEX FOR JOIN",
+				Values: ih.use.forJoin,
+			},
+			{
+				Name:   "USE INDEX FOR ORDER BY",
+				Values: ih.use.forOrderBy,
+			},
+			{
+				Name:   "USE INDEX FOR GROUP BY",
+				Values: ih.use.forGroupBy,
+			},
+		}...)
```

Comment:
I don't want to test your patience, but you could write this in a shorter and faster way:

```go
if ih.use != nil {
    b = appendIndexHints(fmter, b, "USE INDEX", ih.use)
}
if ih.ignore != nil {
    b = appendIndexHints(fmter, b, "IGNORE INDEX", ih.ignore)
}
if ih.ignore != nil {
    b = appendIndexHints(fmter, b, "FORCE", ih.force)
}
```

Then in `appendIndexHints`:

```go
func appendIndexHints(fmter Formatter, b []byte, prefix string, ih *indexHints) {
    b = appendIndexHint(fmter, b, prefix, "", ih.names)
    b = appendIndexHint(fmter, b, prefix, "FOR JOIN", ih.forJoin)
    b = appendIndexHint(fmter, b, prefix, "FOR ORDER BY", ih.forOrderBy)
    b = appendIndexHint(fmter, b, prefix, "FOR GROUP BY", ih.forGroupBy)
    return b
}
```

Would be nice you could send a cleanup PR :+1:

Reply:
Thank you! 
I will send changes in another PR